python/pip_install/private/requirements_parser.bzl

```diff
@@ -0,0 +1,89 @@
+"Pip requirements parser for Starlark"
+
+_STATE = struct(
+    # Consume extraneous whitespace
+    ConsumeSpace = 0,
+    # Consume a comment
+    ConsumeComment = 1,
+    # Parse the name of a pip package
+    ParseDependency = 2,
+    # Parse a full requirement line
+    ParseRequirement = 3,
+)
+
+EOF = {}
+
+def parse(content):
+    """A simplistic (and incomplete) pip requirements lockfile parser.
+
+    Parses the name of a package and its requirement line, but not any
+    other details about the requirement.
+
+    Args:
+      content: lockfile content as a string
+
+    Returns:
+      List of requirements, where each requirement is a 2-element tuple
+      containing the package name and the requirement line.
+
+      E.g., [(certifi', 'certifi==2021.10.8 --hash=sha256:7888...'), ...]
+    """
+    content = content.replace("\r", "")
+
+    requirements = []
+    state = _STATE.ConsumeSpace
+    buffer = ""
+
+    inputs = content.elems()[:]
+    inputs.append(EOF)
+
+    for input in inputs:
+        if state == _STATE.ConsumeSpace:
+            (state, buffer) = _handleConsumeSpace(input)
+        elif state == _STATE.ConsumeComment:
+            (state, buffer) = _handleConsumeComment(input, buffer, requirements)
+        elif state == _STATE.ParseDependency:
+            (state, buffer) = _handleParseDependency(input, buffer, requirements)
+        elif state == _STATE.ParseRequirement:
+            (state, buffer) = _handleParseRequirement(input, buffer, requirements)
+        else:
+            fail("Unknown state %d" % state)
+
+    return requirements
+
+def _handleConsumeSpace(input):
+    if input == EOF:
+        return (_STATE.ConsumeSpace, "")
+    if input.isspace():
+        return (_STATE.ConsumeSpace, "")
+    elif input == "#":
+        return (_STATE.ConsumeComment, "")
+
+    return (_STATE.ParseDependency, input)
+
+def _handleConsumeComment(input, buffer, requirements):
+    if input == "\n":
+        if len(requirements) > 0 and len(requirements[-1]) == 1:
+            return (_STATE.ParseRequirement, buffer)
+        return (_STATE.ConsumeSpace, "")
+    return (_STATE.ConsumeComment, buffer)
+
+def _handleParseDependency(input, buffer, requirements):
+    if input == EOF:
+        fail("Enountered unexpected end of file while parsing requirement")
+    if input.isspace() or input in [">", "<", "~", "=", ";", "["]:
```

Comment:
I'm not sure it is valid for a constraint portion of a requirement string to start with a semi-colon, but you should add a test case for it anyway. e.g `parse("SomePackage;") == ("SomePackage", "SomePackage;")`

Reply:
I thought I saw an example of this somewhere, but I'll double check/add a test case.

Reply:
There are some examples here: https://pip.pypa.io/en/stable/reference/requirement-specifiers/#requirement-specifiers
E.g.
```
SomeProject ; sys_platform == 'win32'
```
Looks like I already do have a test case.